#! /bin/sh /usr/share/dpatch/dpatch-run
## opt-47_sourcecaps.dpatch by Christian Schuld <chris@sonnengesicht.de>
## http://linvdr.org/mailinglists/vdr/2004/08/msg00521.html
## http://article.gmane.org/gmane.linux.vdr/27260/
##
## Thomas Günther <tom@toms-cafe.de>:
##   - adapted to VDR-1.5.0
##   - adapted to VDR-1.7.0
##   - updated to version of extensions patch 72
##   - adapted to VDR-1.7.8
## Tobias Grimm <tg@e-tobi.net>:
##   - adapted to VDR-1.5.16
##
## All lines beginning with `## DP:' are a description of the patch.
## DP: Patch for assigning special satellites to one dvb card.

@DPATCH@
diff -Naurp vdr-1.7.8/config.c vdr-1.7.8-sourcecaps/config.c
--- vdr-1.7.8/config.c	2009-06-13 10:25:05.000000000 +0000
+++ vdr-1.7.8-sourcecaps/config.c	2009-06-15 23:02:05.000000000 +0000
@@ -15,6 +15,7 @@
 #include "interface.h"
 #include "plugin.h"
 #include "recording.h"
+#include "sources.h"
 
 // IMPORTANT NOTE: in the 'sscanf()' calls there is a blank after the '%d'
 // format characters in order to allow any number of blanks after a numeric
@@ -231,6 +232,8 @@ cSetup::cSetup(void)
   LnbFrequLo =  9750;
   LnbFrequHi = 10600;
   DiSEqC = 0;
+  memset(SourceCaps, 0, sizeof SourceCaps);
+  SourceCapsSet = false;
   SetSystemTime = 0;
   TimeSource = 0;
   TimeTransponder = 0;
@@ -400,6 +403,49 @@ bool cSetup::ParseLanguages(const char *
   return true;
 }
 
+void cSetup::StoreSourceCaps(const char *Name)
+{
+  cSetupLine *l;
+  while ((l = Get(Name)) != NULL)
+        Del(l);
+
+  for (int i = 0; i < MAXDEVICES; i++) {
+      char buffer[MAXSOURCECAPS*8]={0,}, *q = buffer;
+      int j = 0;
+      while (SourceCaps[i][j] && j < MAXSOURCECAPS) {
+            if (j==0) 
+               q += snprintf(buffer, sizeof(buffer), "%i ", i+1);
+            q += snprintf(q, sizeof(buffer) - (q-buffer), "%s ", *cSource::ToString(SourceCaps[i][j++]));
+      }
+      if (*buffer)
+         Store(Name, buffer, NULL, true);
+      }
+}
+
+bool cSetup::ParseSourceCaps(const char *Value)
+{
+  char *p;
+  int d = strtol(Value, &p, 10)-1, i = 0;
+  while (p < Value+strlen(Value)) {
+        if (*p==0) return true;
+        if (isblank(*p)) ++p;
+        if (isalpha(*p)) {
+           int source = cSource::FromString(p);
+           if (source != cSource::stNone) {
+              SourceCaps[d][i++] = source;
+              SourceCapsSet = true;
+              }
+           else
+              return false;
+           while (!isblank(*p) && *p)
+                 ++p;
+           if (i>MAXSOURCECAPS)
+              return false;
+           }
+        }
+  return true;
+}
+
 bool cSetup::Parse(const char *Name, const char *Value)
 {
   if      (!strcasecmp(Name, "OSDLanguage"))       { strn0cpy(OSDLanguage, Value, sizeof(OSDLanguage)); I18nSetLocale(OSDLanguage); }
@@ -418,6 +464,7 @@ bool cSetup::Parse(const char *Name, con
   else if (!strcasecmp(Name, "LnbFrequLo"))          LnbFrequLo         = atoi(Value);
   else if (!strcasecmp(Name, "LnbFrequHi"))          LnbFrequHi         = atoi(Value);
   else if (!strcasecmp(Name, "DiSEqC"))              DiSEqC             = atoi(Value);
+  else if (!strcasecmp(Name, "SourceCaps"))          return ParseSourceCaps(Value);
   else if (!strcasecmp(Name, "SetSystemTime"))       SetSystemTime      = atoi(Value);
   else if (!strcasecmp(Name, "TimeSource"))          TimeSource         = cSource::FromString(Value);
   else if (!strcasecmp(Name, "TimeTransponder"))     TimeTransponder    = atoi(Value);
@@ -510,6 +557,7 @@ bool cSetup::Save(void)
   Store("LnbFrequLo",         LnbFrequLo);
   Store("LnbFrequHi",         LnbFrequHi);
   Store("DiSEqC",             DiSEqC);
+  if (SourceCapsSet) StoreSourceCaps("SourceCaps");
   Store("SetSystemTime",      SetSystemTime);
   Store("TimeSource",         cSource::ToString(TimeSource));
   Store("TimeTransponder",    TimeTransponder);
diff -Naurp vdr-1.7.8/config.h vdr-1.7.8-sourcecaps/config.h
--- vdr-1.7.8/config.h	2009-05-21 11:11:32.000000000 +0000
+++ vdr-1.7.8-sourcecaps/config.h	2009-06-15 23:02:05.000000000 +0000
@@ -44,6 +44,9 @@
 #define MINOSDHEIGHT  324
 #define MAXOSDHEIGHT 1200
 
+#define MAXDEVICES         16 // the maximum number of devices in the system
+#define MAXSOURCECAPS     128 // the maximum number of different sources per device
+
 #define MaxFileName 256
 #define MaxSkinName 16
 #define MaxThemeName 16
@@ -192,6 +195,8 @@ private:
   void StoreLanguages(const char *Name, int *Values);
   bool ParseLanguages(const char *Value, int *Values);
   bool Parse(const char *Name, const char *Value);
+  void StoreSourceCaps(const char *Name);
+  bool ParseSourceCaps(const char *Value);
   cSetupLine *Get(const char *Name, const char *Plugin = NULL);
   void Store(const char *Name, const char *Value, const char *Plugin = NULL, bool AllowMultiple = false);
   void Store(const char *Name, int Value, const char *Plugin = NULL);
@@ -215,6 +220,8 @@ public:
   int LnbFrequLo;
   int LnbFrequHi;
   int DiSEqC;
+  int SourceCaps[MAXDEVICES][MAXSOURCECAPS];
+  bool SourceCapsSet;
   int SetSystemTime;
   int TimeSource;
   int TimeTransponder;
diff -Naurp vdr-1.7.8/device.c vdr-1.7.8-sourcecaps/device.c
--- vdr-1.7.8/device.c	2009-06-06 13:25:58.000000000 +0000
+++ vdr-1.7.8-sourcecaps/device.c	2009-06-15 23:02:05.000000000 +0000
@@ -94,8 +94,10 @@ cDevice::cDevice(void)
   for (int i = 0; i < MAXRECEIVERS; i++)
       receiver[i] = NULL;
 
-  if (numDevices < MAXDEVICES)
+  if (numDevices < MAXDEVICES) {
      device[numDevices++] = this;
+     SetSourceCaps(cardIndex);
+     }
   else
      esyslog("ERROR: too many devices!");
 }
@@ -313,6 +315,17 @@ void cDevice::SetCamSlot(cCamSlot *CamSl
   camSlot = CamSlot;
 }
 
+void cDevice::SetSourceCaps(int Index)
+{
+  for (int d = 0; d < numDevices; d++) {
+      if (Index < 0 || Index == device[d]->CardIndex()) {
+         for (int i = 0; i < MAXSOURCECAPS; i++)
+             device[d]->sourceCaps[i] = Setup.SourceCaps[device[d]->CardIndex()][i];
+         }
+      }
+}
+
+
 void cDevice::Shutdown(void)
 {
   primaryDevice = NULL;
diff -Naurp vdr-1.7.8/device.h vdr-1.7.8-sourcecaps/device.h
--- vdr-1.7.8/device.h	2009-06-06 11:15:49.000000000 +0000
+++ vdr-1.7.8-sourcecaps/device.h	2009-06-15 23:02:05.000000000 +0000
@@ -25,7 +25,6 @@
 #include "thread.h"
 #include "tools.h"
 
-#define MAXDEVICES         16 // the maximum number of devices in the system
 #define MAXPIDHANDLES      64 // the maximum number of different PIDs per device
 #define MAXRECEIVERS       16 // the maximum number of receivers per device
 #define MAXVOLUME         255
@@ -145,6 +144,8 @@ public:
   static void SetAvoidDevice(cDevice *Device) { avoidDevice = Device; }
          ///< Sets the given Device to be temporarily avoided in the next call to
          ///< GetDevice(const cChannel, int, bool).
+  static void SetSourceCaps(int Index = -1);
+         ///< Sets the SourceCaps of the given device according to the Setup data.
   static void Shutdown(void);
          ///< Closes down all devices.
          ///< Must be called at the end of the program.
@@ -152,6 +153,7 @@ private:
   static int nextCardIndex;
   int cardIndex;
 protected:
+  int sourceCaps[MAXSOURCECAPS];
   cDevice(void);
   virtual ~cDevice();
   virtual bool Ready(void);
diff -Naurp vdr-1.7.8/dvbdevice.c vdr-1.7.8-sourcecaps/dvbdevice.c
--- vdr-1.7.8/dvbdevice.c	2009-06-06 11:17:20.000000000 +0000
+++ vdr-1.7.8-sourcecaps/dvbdevice.c	2009-06-15 23:20:35.000000000 +0000
@@ -901,6 +901,12 @@ void cDvbDevice::TurnOffLiveMode(bool Li
 bool cDvbDevice::ProvidesSource(int Source) const
 {
   int type = Source & cSource::st_Mask;
+  if (Setup.SourceCapsSet && type == cSource::stSat && (frontendType == SYS_DVBS || frontendType == SYS_DVBS2)) {
+     for (int i = 0; i < MAXSOURCECAPS; i++)
+         if (sourceCaps[i] == Source)
+            return true;
+     return false;
+     }
   return type == cSource::stNone
       || type == cSource::stCable && (frontendType == SYS_DVBC_ANNEX_AC || frontendType == SYS_DVBC_ANNEX_B)
       || type == cSource::stSat   && (frontendType == SYS_DVBS || frontendType == SYS_DVBS2)
diff -Naurp vdr-1.7.8/sources.c vdr-1.7.8-sourcecaps/sources.c
--- vdr-1.7.8/sources.c	2008-02-10 14:07:26.000000000 +0000
+++ vdr-1.7.8-sourcecaps/sources.c	2009-06-15 23:02:05.000000000 +0000
@@ -68,7 +68,7 @@ int cSource::FromString(const char *s)
      int pos = 0;
      bool dot = false;
      bool neg = false;
-     while (*++s) {
+     while (*++s && !isblank(*s)) {
            switch (toupper(*s)) {
              case '0' ... '9': pos *= 10;
                                pos += *s - '0';
